src/traces/bar/calc.js

```diff
@@ -51,15 +51,21 @@ module.exports = function calc(gd, trace) {
     var serieslen = Math.min(pos.length, size.length),
         cd = [];
 
-    // set position
+    // set position and size
     for(i = 0; i < serieslen; i++) {
+        var p = pos[i];
+        var s = size[i];
 
         // add bars with non-numeric sizes to calcdata
         // so that ensure that traces with gaps are
         // plotted in the correct order
 
-        if(isNumeric(pos[i])) {
-            cd.push({p: pos[i]});
+        if(isNumeric(p)) {
+            if(isNumeric(s)) {
+                cd.push({ p: p, s: s });
+            } else {
+                cd.push({ p: p });
+            }
```

Comment:
OK, but then how do you want to solve this for bar color arrays etc? save the original index in each `cd` element? `cd.push({p: p, s: s, i: i})` or something? That would need to be propagated all the way through to `Lib.mergeArray` which seems like a bit of a disaster.

Non-numerics should be regarded as an edge case, so the overhead of managing their objects shouldn't be seen as a concern, but you're right to worry about the overhead of checking each *valid* point. That said, this works for scatter which can have way more points than bars, so I think as long as we make this test super simple it shouldn't be a problem.